### PR TITLE
Get pin manager working

### PR DESCRIPTION
**Pull request by rockboynton**
# Date

Initial PR Date: 3/21/21

Due Date: 3/26/21

# Reviewers
 @Ivendrakeo @anthony-bartman 

# Description

See commit messages for details

# Project Issue

#38 

# Test Results

Testing needed @iliescua @DeclanBruce 
Make sure you run the init script before attempting to test

### CODE REVIEW

**Review by anthony-bartman — COMMENTED** [review]
I'm sure you talked with Draven about changing the program manager path to '/home/myprograms'.  All looks good to me! Looks you're adding the pin wrapper, an uptodate pin-manager.py that I will use, adding our library, and a few tweaks here or there to make our program nicer. Looks solid to me

**Comment by Ivendrakeo** [comment]
> I'm sure you talked with Draven about changing the program manager path to '/home/myprograms'. All looks good to me! Looks you're adding the pin wrapper, an uptodate pin-manager.py that I will use, adding our library, and a few tweaks here or there to make our program nicer. Looks solid to me

why would we be changing the program manager to `/home/myporgrams`? `/etc/hobby-hub/` is where we wanted all of our source code to be located. and I'd say the program manager is source code. `/home/` usually contains linux user accounts only. ie debian & dev.

**Comment by rockboynton** [comment]
> > I'm sure you talked with Draven about changing the program manager path to '/home/myprograms'. All looks good to me! Looks you're adding the pin wrapper, an uptodate pin-manager.py that I will use, adding our library, and a few tweaks here or there to make our program nicer. Looks solid to me
> 
> why would we be changing the program manager to `/home/myporgrams`? `/etc/hobby-hub/` is where we wanted all of our source code to be located. and I'd say the program manager is source code. `/home/` usually contains linux user accounts only. ie debian & dev.

That's not what is happening

File: files/init/start.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # act as an entry point to startup the rest of the hobby-hub utilities (as needed).
 
 # Start program manager
-sudo python3 /etc/hobby-hub/program_manager.py /etc/hobby-hub/test_programs/ > /etc/hobby-hub/out/running_programs.txt
+sudo python3 /etc/hobby-hub/program_manager.py /home/dev/myprograms/ > /etc/hobby-hub/out/running_programs.txt
```
Comment (rockboynton) [inline files/init/start.sh]: @anthony-bartman @Ivendrakeo This just allows the users to write their programs in the home directory, rather than running the test programs from the test directory

Comment (Ivendrakeo) [reply files/init/start.sh]: so the way I had things setup with the cmd line script was to use /etc/hobby-hub/test_programs as the program directory. then they would develop in /home/dev/myprograms and transfer then to /etc/hobby-hub/tes_programs using the utility and that would guarntee they would be run on startup. 

Think about it. your using /etc/home/dev/ as your workspace and what if your crafting a program that you don't want to run? you can still use the cmd-line utility to debug from your workspace, but when you acutally want it to run on startup, you should use the utility to move it to the test_programs dir. 

we could change the name of test_programs because in this case it's acting more as a staging directory but the user doesn't actually see the name of the dir so i didn't bother changing it thus far.

In short, I vote to veto this change.

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            logging.info(f'{tag} added to JSON with placeholder')
+
+        return physical_pin
+
+
+def get_gpio(tag):
+    return Pin(request_pin(tag, Pin_t.GPIO))
+
+
+def get_led(tag):
```
Comment (Ivendrakeo) [inline files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:294]: could you add some documentation to the methods the user should be using? in the code and probobly a usage guide on the wiki?

File: files/src/hobby_hub_utils/test/test_pin.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from hobby_hub_pkg.pin_manager import get_led
+import unittest
+import time
+
+class TestPinManager(unittest.TestCase):
+
+    def test_blink_led0(self):
+        led = get_led("myled0")
```
Comment (Ivendrakeo) [inline files/src/hobby_hub_utils/test/test_pin.py:8]: what about some of the other methods? I don't think get_gpio would be different but i'd appreciate a test case for at least this as well.

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+def reset(pin_config_filename):
+    """Resets the pin configuration JSON file"""
+    logging.info('Resetting config')
+    with open(pin_config_filename, 'w') as f:
+        f.write("{}")
+        logging.info('Config reset')
+
+
+def clear_unused(pin_config_filename):
+    """Identify and remove unused key-value pairs
```
Comment (Ivendrakeo) [inline files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:245]: so when is the user supposed to be calling this? if it can be called on startup then maybe it would be wise to make a python script that runs on startup and have it call this? because I agree, there should be some way of auto-clearing unused pins.

File: files/src/hobby_hub_utils/test/test_pin.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from hobby_hub_pkg.pin_manager import get_led
+import unittest
+import time
+
+class TestPinManager(unittest.TestCase):
```
Comment (Ivendrakeo) [inline files/src/hobby_hub_utils/test/test_pin.py:5]: also a test where you request and use a pin in one file then request the same pin in another to view the state would be a useful test.  

from my perspective, it should be able to view the state in the second program as it changes in the first.

File: files/src/program_manager.py [hunk 6]
```diff
@@ -44,8 +44,8 @@ def main(arguments):
         pid = Popen(["python3", program]).pid
         programs[program] = pid
 
-    if c_files or cpp_files:
-        print('Compiling source files...')
+    #if c_files or cpp_files:
```
Comment (Ivendrakeo) [inline files/src/program_manager.py]: can you remove these changes to the program manager? this is a relic from when I was working on this branch... alternatively I can revert these.

**Review by Ivendrakeo — CHANGES_REQUESTED** [review]

**Comment by Ivendrakeo** [comment]
Also I checked out your branch, ran the setup script and rebooted the board. when i logged back in, the /etc/hobby-hub/pin_mapping.json was unedited. isn't that supposed to change? also i'm getting 2 blinking led's but they are blinking very erratically and much quicker with less opacity. Not the same as the previous test programs.

File: files/init/start.sh (hunk 1 above)
Comment (rockboynton) [reply files/init/start.sh]: Okay I didnt know thats how you set it up, we should change the name to just `/etc/hobby-hub/programs` though

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py (hunk 4 above)
Comment (rockboynton) [reply files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:245]: Yeah I just had it in there cause you said it should be there, but there is currently nothing that defines when a pin is "in use" nor when it would go out of use ... thoughts?

File: files/src/program_manager.py (hunk 6 above)
Comment (rockboynton) [reply files/src/program_manager.py]: Yeah, you can do it I wouldn't mind

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py (hunk 2 above)
Comment (rockboynton) [reply files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:294]: > Also I checked out your branch, ran the setup script and rebooted the board. when i logged back in, the /etc/hobby-hub/pin_mapping.json was unedited. isn't that supposed to change? also i'm getting 2 blinking led's but they are blinking very erratically and much quicker with less opacity. Not the same as the previous test programs.

Change to what? Should the json file be cleared on startup?

File: files/src/hobby_hub_utils/test/test_pin.py (hunk 5 above)
Comment (rockboynton) [reply files/src/hobby_hub_utils/test/test_pin.py:5]: Sure, although I don't see why they have to be from separate files? If you request a pin and then request the same pin in the same file, shouldn't the behavior be the same?

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py (hunk 2 above)
Comment (rockboynton) [reply files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:294]: > could you add some documentation to the methods the user should be using? in the code and probably a usage guide on the wiki?

Sure, shall @DeclanBruce or @iliescua like to test?

File: files/src/hobby_hub_utils/test/test_pin.py (hunk 3 above)
Comment (rockboynton) [reply files/src/hobby_hub_utils/test/test_pin.py:8]: Sure, @DeclanBruce  @iliescua

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py (hunk 2 above)
Comment (Ivendrakeo) [reply files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:294]: doesn't the JSON reflect the current pins in use? since that's what barty is reading.

File: files/src/hobby_hub_utils/test/test_pin.py (hunk 5 above)
Comment (Ivendrakeo) [reply files/src/hobby_hub_utils/test/test_pin.py:5]: well i guess the idea is that you have a common pin between two programs. so having two files kinda simulates that.

**Comment by anthony-bartman** [comment]
Okay, so what's left to chat about? I think you 2 resolved most of the conflicts? Should we wait to chat about merging this branch on Friday ? @rockboynton  @Ivendrakeo

**Comment by rockboynton** [comment]
> Okay, so what's left to chat about? I think you 2 resolved most of the conflicts? Should we wait to chat about merging this branch on Friday ? @rockboynton  @Ivendrakeo 

Did you see my replies? There's still a lot to do

File: files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py (hunk 4 above)
Comment (Ivendrakeo) [reply files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py:245]: it should go out of use when it's not being used by a currently running program. idc how you do this, but to me this could be done by having a script run on startup and look at what pins are in use then remove those which are not. alternatively, you could create a script that runs when you call the "remove program" from the command line utility. (This might actually be easier)

File: files/init/start.sh (hunk 1 above)
Comment (Ivendrakeo) [reply files/init/start.sh]: ok. for now lets NOT make this change until both of our branches are merged in. Afterwards we can make a quick PR changing it. I'll make a git issue for this.

File: files/src/hobby_hub_utils/test/test_pin.py (hunk 3 above)
Comment (rockboynton) [reply files/src/hobby_hub_utils/test/test_pin.py:8]: @anthony-bartman
